signer/src/transaction_signer.rs

```diff
@@ -846,6 +839,90 @@ where
     }
 }
 
+/// Asserts whether a `DkgBegin` WSTS message should be allowed to proceed
+/// based on the current state of the signer and the DKG configuration.
+async fn assert_allow_dkg_begin(
+    context: &impl Context,
+    bitcoin_chain_tip: &model::BitcoinBlockHash,
+) -> Result<(), Error> {
+    let storage = context.get_storage();
+    let config = context.config();
+
+    // Get the bitcoin block at the chain tip so that we know the height
+    let bitcoin_chain_tip_block = storage
+        .get_bitcoin_block(bitcoin_chain_tip)
+        .await?
+        .ok_or(Error::NoChainTip)?;
+
+    // Get the number of DKG shares that have been stored
+    let dkg_shares_entry_count = storage.get_encrypted_dkg_shares_count().await?;
+
+    // Get DKG configuration parameters
+    let dkg_min_bitcoin_block_height = config.signer.dkg_min_bitcoin_block_height;
+    let dkg_target_rounds = config.signer.dkg_target_rounds;
+
+    // Determine the action based on the DKG shares count and the rerun height (if configured)
+    match (
+        dkg_shares_entry_count,
+        dkg_target_rounds,
+        dkg_min_bitcoin_block_height,
+    ) {
+        (0, _, _) => {
+            tracing::info!(
+                ?dkg_min_bitcoin_block_height,
+                %dkg_target_rounds,
+                "no DKG shares exist; proceeding with DKG"
+            );
+        }
+        (current, target, Some(dkg_min_height)) => {
+            if current >= target.get() {
+                tracing::warn!(
+                    ?dkg_min_bitcoin_block_height,
+                    %dkg_target_rounds,
+                    dkg_current_rounds = %dkg_shares_entry_count,
+                    "The target number of DKG shares has been reached; aborting"
+                );
+                return Err(Error::DkgHasAlreadyRun);
+            }
+            if bitcoin_chain_tip_block.block_height < dkg_min_height.get() {
+                tracing::warn!(
+                    ?dkg_min_bitcoin_block_height,
+                    %dkg_target_rounds,
+                    dkg_current_rounds = %dkg_shares_entry_count,
+                    "bitcoin chain tip is below the minimum height for DKG rerun; aborting"
+                );
+                return Err(Error::DkgHasAlreadyRun);
+            }
+            tracing::info!(
+                ?dkg_min_bitcoin_block_height,
+                %dkg_target_rounds,
+                dkg_current_rounds = %dkg_shares_entry_count,
+                "DKG rerun height has been met and we are below the target number of rounds; proceeding with DKG"
+            );
+        }
+        (1, _, None) => {
```

Comment:
qq: why the `1` here? I guess it's related to having those last two cases as separate ones; is it just for better conf debugging?

Reply:
Actually a miss -- updated in [ec63981](https://github.com/stacks-network/sbtc/pull/1205/commits/ec63981d4b083be5819e8bfc47c1b78d392b584e)